core/src/test/scala/unit/kafka/coordinator/group/PartitionWriterImplTest.scala

```diff
@@ -0,0 +1,225 @@
+      new ListenerAdaptor(listener)
+    )
+    assertEquals(
+      new ListenerAdaptor(listener).hashCode(),
+      new ListenerAdaptor(listener).hashCode()
+    )
+  }
+
+  @Test
+  def testWriteRecords(): Unit = {
+    val tp = new TopicPartition("foo", 0)
+    val replicaManager = mock(classOf[ReplicaManager])
+    val time = new MockTime()
+    val partitionRecordWriter = new PartitionWriterImpl(
+      replicaManager,
+      new StringKeyValueSerializer(),
+      CompressionType.NONE,
+      time
+    )
+
+    when(replicaManager.getLogConfig(tp)).thenReturn(Some(LogConfig.fromProps(
+      Collections.emptyMap(),
+      new Properties()
+    )))
+
+    val recordsCapture: ArgumentCaptor[Map[TopicPartition, MemoryRecords]] =
+      ArgumentCaptor.forClass(classOf[Map[TopicPartition, MemoryRecords]])
+    when(replicaManager.appendToLocalLog(
+      ArgumentMatchers.eq(true),
+      ArgumentMatchers.eq(AppendOrigin.COORDINATOR),
+      recordsCapture.capture(),
+      ArgumentMatchers.eq(1),
+      ArgumentMatchers.eq(RequestLocal.NoCaching)
+    )).thenReturn(Map(tp -> LogAppendResult(new LogAppendInfo(
+      Optional.empty(),
+      10,
+      OptionalInt.empty(),
+      RecordBatch.NO_TIMESTAMP,
+      -1L,
+      RecordBatch.NO_TIMESTAMP,
+      -1L,
+      RecordConversionStats.EMPTY,
+      CompressionType.NONE,
+      CompressionType.NONE,
+      -1,
+      -1,
+      false,
+      -1L,
+      Collections.emptyList(),
+      "",
+      LeaderHwChange.INCREASED
+    ))))
+
+    val records = List(
+      ("k0", "v0"),
+      ("k1", "v1"),
+      ("k2", "v2"),
+    )
+
+    assertEquals(11, partitionRecordWriter.append(
+      tp,
+      records.asJava
+    ))
+
+    verify(replicaManager).maybeCompletePurgatories(
+      tp,
+      LeaderHwChange.INCREASED
+    )
+
+    val batch = recordsCapture.getValue.getOrElse(tp,
+      throw new AssertionError(s"No records for $tp"))
+    assertEquals(1, batch.batches().asScala.toList.size)
+
+    val receivedRecords = batch.records.asScala.map { record =>
+      (
+        Charset.defaultCharset().decode(record.key).toString,
+        Charset.defaultCharset().decode(record.value).toString,
+      )
+    }.toList
+
+    assertEquals(records, receivedRecords)
+  }
+
+  @Test
+  def testWriteRecordTooLarge(): Unit = {
+    val tp = new TopicPartition("foo", 0)
+    val replicaManager = mock(classOf[ReplicaManager])
+    val partitionRecordWriter = new PartitionWriterImpl(
+      replicaManager,
+      new StringKeyValueSerializer(),
+      CompressionType.NONE,
+      Time.SYSTEM
+    )
+
+    val maxBatchSize = 16384
+    when(replicaManager.getLogConfig(tp)).thenReturn(Some(LogConfig.fromProps(
+      Map(TopicConfig.MAX_MESSAGE_BYTES_CONFIG -> maxBatchSize).asJava,
+      new Properties()
+    )))
+
+    val randomBytes = TestUtils.randomBytes(maxBatchSize + 1)
+    // We need more than one record here because the first record
+    // is always allowed by the MemoryRecordsBuilder.
```

Comment:
Is this a bug/quirk or is there a reason that the first record is always allowed?